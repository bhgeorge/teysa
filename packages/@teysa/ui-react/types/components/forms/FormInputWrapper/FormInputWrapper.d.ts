import { ReactNode } from 'react';
import { IconType } from '../../atomic/Icon/Icon';
import { FormInputBase } from '../form-input-base';
interface FormInputProps extends FormInputBase {
    children?: ReactNode | ReactNode[];
    icon?: IconType;
    [name: string]: unknown;
}
declare const FormInputWrapper: ({ children, label, error, helpText, name, icon, required, ...props }: FormInputProps) => JSX.Element;
export default FormInputWrapper;
