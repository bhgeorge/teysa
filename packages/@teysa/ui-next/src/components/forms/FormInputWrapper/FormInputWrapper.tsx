// Global
import * as React from 'react';
import * as Formik from 'formik';
// Components
import { Icon } from '../../atomic/Icon/Icon';
// Local
import {
  FormInputBase,
  getDescribedBy,
  getErrorId,
  getHelpTextId,
  getInputClasses,
} from '../form-input-base';

type FormInputWrapperProps = FormInputBase &
  React.PropsWithChildren & {
    [name: string]: unknown;
  };

export function FormInputWrapper({
  children,
  label,
  helpText,
  name,
  icon,
  required,
  ...props
}: FormInputWrapperProps) {
  const id = React.useId();

  const { errors, touched } = Formik.useFormikContext();

  const hasError = Boolean(
    (touched as Record<string, boolean | undefined>)[name] &&
      (errors as Record<string, string | undefined>)[name]
  );

  return (
    <div>
      <label htmlFor={id} className="block mb-1 text-sm">
        {label}
      </label>
      <div className="relative">
        <Formik.Field
          aria-describedby={getDescribedBy(id, hasError, helpText)}
          aria-required={required}
          className={getInputClasses(hasError)}
          id={id}
          name={name}
          {...props}
        >
          {children}
        </Formik.Field>
        {icon && (
          <span className="absolute top-1/2 right-1 -translate-y-1/2 pointer-events-none">
            <Icon icon={icon} />
          </span>
        )}
      </div>
      <Formik.ErrorMessage name={name}>
        {err => (
          <p role="alert" id={getErrorId(id)} className="text-theme-error my-2 text-sm">
            {err}
          </p>
        )}
      </Formik.ErrorMessage>
      {helpText && (
        <p id={getHelpTextId(id)} className="text-theme-text-alt my-2 text-sm">
          {helpText}
        </p>
      )}
    </div>
  );
}
