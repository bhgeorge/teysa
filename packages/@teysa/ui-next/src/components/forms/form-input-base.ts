// Global
import classNames from 'classnames';
// Components
import { IconType } from '../atomic/Icon/Icon';

export interface FormInputBase {
  error?: string;
  helpText?: string;
  icon?: IconType;
  label: string;
  name: string;
  required?: boolean;
}

// Styles
export const getInputClasses = (hasError: boolean) => {
  const borderColor = hasError ? 'border-theme-error' : 'border-theme-bg-alt';
  return classNames(
    'appearance-none',
    'bg-theme-bg-alt',
    'border',
    'px-1',
    'py-1.5',
    'relative',
    'w-full',
    borderColor
  );
};

// ID helpers
export const getErrorId = (id: string): string => `${id}__error`;
export const getHelpTextId = (id: string): string => `${id}__help-text`;
export const getDescribedBy = (
  id: string,
  error?: string,
  helpText?: string
): string | null => {
  const describedBy = [];

  if (error) {
    describedBy.push(getErrorId(id));
  }
  if (helpText) {
    describedBy.push(getHelpTextId(id));
  }

  if (describedBy.length === 0) {
    return null;
  }

  return describedBy.join(' ');
};
