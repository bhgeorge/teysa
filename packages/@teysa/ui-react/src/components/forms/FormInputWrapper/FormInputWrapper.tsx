// Global
import React, { useId, ReactNode } from 'react';
import { Field } from 'formik';
// Local
import {
  FormInputBase,
  getDescribedBy,
  getErrorId,
  getHelpTextId,
  getInputClasses,
} from '../form-input-base';

interface FormInputProps extends FormInputBase {
  children?: ReactNode | ReactNode[];
  [name: string]: unknown;
}

const FormInputWrapper = ({
  children,
  label,
  error,
  helpText,
  name,
  required,
  ...props
}: FormInputProps) => {
  const id = useId();
  return (
    <div>
      <label htmlFor={id} className="block mb-1 text-sm">
        {label}
      </label>
      <Field
        aria-describedby={getDescribedBy(id, error, helpText)}
        aria-required={required}
        className={getInputClasses(!!error)}
        id={id}
        name={name}
        {...props}
      >
        {children}
      </Field>
      {error && (
        <p role="alert" id={getErrorId(id)} className="text-theme-error my-2 text-sm">
          {error}
        </p>
      )}
      {helpText && (
        <p id={getHelpTextId(id)} className="text-theme-text-alt my-2 text-sm">
          {helpText}
        </p>
      )}
    </div>
  );
};

export default FormInputWrapper;
