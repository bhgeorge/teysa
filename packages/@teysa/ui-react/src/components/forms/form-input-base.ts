import classNames from 'classnames';

export interface FormInputBase {
  error?: string;
  helpText?: string;
  label: string;
  name: string;
  required?: boolean;
}

// Styles
export const getInputClasses = (hasError: boolean) => {
  const borderColor = hasError ? 'border-theme-error' : 'border-theme-bg-alt';
  return classNames('bg-theme-bg-alt', 'border', 'w-full', 'py-1.5', 'px-1', borderColor);
};

// ID helpers
export const getErrorId = (id: string): string => `${id}__error`;
export const getHelpTextId = (id: string): string => `${id}__help-text`;
export const getDescribedBy = (id: string, error?: string, helpText?: string): string | null => {
  const describedBy = [];

  if (error) {
    describedBy.push(getErrorId(id));
  }
  if (helpText) {
    describedBy.push(getHelpTextId(id));
  }

  if (describedBy.length === 0) {
    return null;
  }

  return describedBy.join(' ');
};
