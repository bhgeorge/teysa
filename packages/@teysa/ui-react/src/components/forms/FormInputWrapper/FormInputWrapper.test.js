// Global
import { render } from '@testing-library/react';
import React from 'react';
import { Formik } from 'formik';
// Local
import FormInputWrapper from './FormInputWrapper';
import { getErrorId, getHelpTextId } from '../form-input-base';

const FIELD_NAME = 'firstName';
const LABEL_TEXT = 'First name';

const DEFAULT_PROPS = {
  label: LABEL_TEXT,
  name: FIELD_NAME,
};

const createComponent = (overrideProps = {}) => {
  const props = {
    ...DEFAULT_PROPS,
    ...overrideProps,
  };

  return render(
    <Formik onSubmit={() => {}} initialValues={{}}>
      <FormInputWrapper {...props} />
    </Formik>
  );
};

describe('in any state', () => {
  it('displays the input label', () => {
    const component = createComponent();

    // We contain the correct "for" attribute
    expect(component.queryByLabelText(LABEL_TEXT)).not.toBe(null);
  });
});

describe('if there is an error', () => {
  it('displays an error', () => {
    const ERROR_TEXT = 'First name is required';
    const component = createComponent({ error: ERROR_TEXT });
    const error = component.getByText(ERROR_TEXT);
    const id = component.getByRole('textbox').getAttribute('id');
    expect(error.getAttribute('role')).toBe('alert');
    expect(error.getAttribute('id')).toBe(getErrorId(id));
  });
});

describe('if there is help text', () => {
  it('displays the help text', () => {
    const HELP_TEXT = 'Passwords must be 12 characters in length.';
    const component = createComponent({ helpText: HELP_TEXT });
    const helpText = component.getByText(HELP_TEXT);
    const id = component.getByRole('textbox').getAttribute('id');
    expect(helpText.getAttribute('id')).toBe(getHelpTextId(id));
  });
});
