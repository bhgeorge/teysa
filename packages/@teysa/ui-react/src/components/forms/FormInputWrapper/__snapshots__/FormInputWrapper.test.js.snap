// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if there is an icon renders the icon 1`] = `
<div>
  <label
    class="block mb-1 text-sm"
    for=":r3:"
  >
    First name
  </label>
  <div
    class="relative"
  >
    <input
      class="appearance-none bg-theme-bg-alt border px-1 py-1.5 relative w-full border-theme-bg-alt"
      id=":r3:"
      name="firstName"
      value=""
    />
    <span
      class="absolute top-1/2 right-1 -translate-y-1/2 pointer-events-none"
    >
      <svg
        class="h-6 w-6"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
          fill="currentColor"
        />
      </svg>
    </span>
  </div>
</div>
`;
